tests/TestCase/Database/ExpressionTypeCastingIntegrationTest.php

```diff
@@ -0,0 +1,209 @@
+    {
+        $substr = function ($start, $lenght = null) use ($value) {
+            return new FunctionExpression(
+                'SUBSTR',
+                $lenght === null ? [$value, $start] : [$value, $start, $lenght],
+                ['string', 'integer', 'integer']
+            );
+        };
+        return new FunctionExpression(
+            'CONCAT',
+            [$substr(15, 4), $substr(10, 4), $substr(1, 8), $substr(20, 4), $substr(25)]
+        );
+    }
+}
+
+/**
+ * Tests for Expression objects casting values to other expressions
+ * using the type classes
+ *
+ */
+class ExpressionTypeCastingIntegrationTest extends TestCase
+{
+
+    public $fixtures = ['core.ordered_uuid_items'];
+
+    public function setUp()
+    {
+        parent::setUp();
+        $this->connection = ConnectionManager::get('test');
+        $this->skipIf($this->connection->driver() instanceof Sqlserver, 'This tests uses functions specific to other drivers');
+        Type::map('ordered_uuid', OrderedUuidType::class);
+    }
+
+    protected function _insert()
+    {
+        $query = $this->connection->newQuery();
+        $values = $query
+            ->insert(['id', 'published', 'name'], ['id' => 'ordered_uuid'])
+            ->into('ordered_uuid_items')
+            ->clause('values');
+        $values
+            ->values([
+                ['id' => '481fc6d0-b920-43e0-a40d-6d1740cf8569', 'published' => 0, 'name' => 'Item 1'],
+                ['id' => '48298a29-81c0-4c26-a7fb-413140cf8569', 'published' => 0, 'name' => 'Item 2'],
+                ['id' => '482b7756-8da0-419a-b21f-27da40cf8569', 'published' => 0, 'name' => 'Item 3'],
+            ]);
+
+        $query->execute();
+    }
+
+    /**
+     * Tests inserting a value that is to be converted to an expression
+     * automatically
+     *
+     * @return void
+     */
+    public function testInsert()
+    {
+        $this->_insert();
+        $query = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->order('id')
+            ->defaultTypes(['id' => 'ordered_uuid']);
+
+        $query->selectTypeMap($query->typeMap());
+        $results = $query->execute()->fetchAll('assoc');
+
+        $this->assertEquals(new UuidValue('419a8da0482b7756b21f27da40cf8569'), $results[0]['id']);
+        $this->assertEquals(new UuidValue('43e0b920481fc6d0a40d6d1740cf8569'), $results[1]['id']);
+        $this->assertEquals(new UuidValue('4c2681c048298a29a7fb413140cf8569'), $results[2]['id']);
+    }
+
+    /**
+     * Test selecting with a custom expression type using conditions
+     *
+     * @return void
+     */
+    public function testSelectWithConditions()
+    {
+        $this->_insert();
+        $result = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->where(['id' => '48298a29-81c0-4c26-a7fb-413140cf8569'], ['id' => 'ordered_uuid'])
```

Comment:
Yes, but that will require special code in the Type class... not sure that is interesting to see, as it has been shown many times for example with the DateTimeType